### PR TITLE
KAZOO-4528: lock whapps_config if parameter sets in config file

### PR DESCRIPTION
**Pull request by icehess**

### CODE REVIEW

**Comment by lazedo** [comment]
@icehess i would simply add a `maybe_save_category/5` with 2 function clauses and pass `lock_whapps_config` from `maybe_save_category/4`

**Comment by icehess** [comment]
I just wanted to avoid unnecessary `couch_mgr:open_cache_doc` in `update_category/5`

**Comment by lazedo** [comment]
@icehess we don't want to lock the `read`, just the `write`

File: core/whistle_config-1.0.0/src/whapps_config.erl [hunk 1]
```diff
@@ -401,10 +401,20 @@ update_category(Category, JObj, PvtFields) ->
...
                                  {'ok', wh_json:object()} |
                                  {'error', 'conflict'}.
+-spec maybe_save_category(ne_binary(), wh_json:object(), api_object(), boolean(), any()) ->
+                                 {'ok', wh_json:object()} |
+                                 {'error', 'conflict'}.
 maybe_save_category(Category, JObj, PvtFields) ->
     maybe_save_category(Category, JObj, PvtFields, 'false').
 
 maybe_save_category(Category, JObj, PvtFields, Looped) ->
+    IsLocked =  wh_config:get_atom('whistle_apps', 'lock_whapps_config', ['false']),
```
Comment (lazedo) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: i would prefer `[IsLocked] = `

File: core/whistle_config-1.0.0/src/whapps_config.erl [hunk 2]
```diff
@@ -401,10 +401,20 @@ update_category(Category, JObj, PvtFields) ->
...
+    maybe_save_category(Category, Keys, Value, Node, Options, IsLocked).
+
+maybe_save_category(_, _, _, _, ['true']) ->
```
Comment (lazedo) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: `true' instead of ['true']

Comment (lazedo) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: also accept JObj (2nd param)

File: core/whistle_config-1.0.0/src/whapps_config.erl [hunk 3]
```diff
@@ -401,10 +401,20 @@ update_category(Category, JObj, PvtFields) ->
...
+    lager:debug("failed to update category, system config doc is locked!"),
+    {'ok', 'doc_locked'};
```
Comment (lazedo) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: and return JObj here

File: core/whistle_config-1.0.0/src/whapps_config.erl [hunk 4]
```diff
@@ -401,10 +401,20 @@ update_category(Category, JObj, PvtFields) ->
...
                                  {'error', 'conflict'}.
+-spec maybe_save_category(ne_binary(), wh_json:object(), api_object(), boolean(), any()) ->
+                                 {'ok', wh_json:object()} |
+                                 {'error', 'conflict'}.
 maybe_save_category(Category, JObj, PvtFields) ->
     maybe_save_category(Category, JObj, PvtFields, 'false').
 
 maybe_save_category(Category, JObj, PvtFields, Looped) ->
+    IsLocked =  wh_config:get_atom('whistle_apps', 'lock_whapps_config', ['false']),
+    maybe_save_category(Category, Keys, Value, Node, Options, IsLocked).
```
Comment (lazedo) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: i believe you want this `maybe_save_category(Category, JObj, PvtFields, Looped, IsLocked)`

Comment (icehess) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: I fixed this.

File: core/whistle_config-1.0.0/src/whapps_config.erl (hunk 3 above)
Comment (icehess) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: what about doing a `couch_mgr:add_to_doc_cache` before returning? Does it cause inconsistency?

File: core/whistle_config-1.0.0/src/whapps_config.erl [hunk 5]
```diff
@@ -401,10 +401,20 @@ update_category(Category, JObj, PvtFields) ->
...
+                                 {'error', 'conflict'}.
 maybe_save_category(Category, JObj, PvtFields) ->
     maybe_save_category(Category, JObj, PvtFields, 'false').
 
 maybe_save_category(Category, JObj, PvtFields, Looped) ->
+    [IsLocked] =  wh_config:get_atom('whistle_apps', 'lock_whapps_config', ['false']),
+    maybe_save_category(Category, JObj, PvtFields, Looped, IsLocked).
+
+maybe_save_category(_, JObj, _, _, 'true') ->
+    lager:debug("failed to update category, system config doc is locked!"),
```
Comment (lazedo) [inline core/whistle_config-1.0.0/src/whapps_config.erl]: @icehess didn't you want to add the `couch_mgr:add_to_doc_cache(?WH_CONFIG_DB, Category, JObj),` here ?
i think it makes sense.

**Comment by fenollp** [comment]
LGTM

**Comment by fogeytron** [comment]
this commit has nothing to do with what this ticket is requesting

**Comment by lazedo** [comment]
@fogeytron can you explain why?

**Comment by jamesaimonetti** [comment]
@lazedo @icehess should have used [KAZOO-4528](https://2600hz.atlassian.net/browse/KAZOO-4528)  but the last commit was [KAZOO-4538](https://2600hz.atlassian.net/browse/KAZOO-4538), which seems to be what JIRA used when the PR was merged.

**Comment by jamesaimonetti** [comment]
@icehess I think the PR is incomplete, by the way. I added a comment to the ticket but there should be a SUP command for toggling the locked state at runtime.

**Comment by lazedo** [comment]
@jamesaimonetti yep, the commit had (wrong) KAZOO-4538 message.

**Pull request merged**